doc/dev/schema.md

```diff
@@ -0,0 +1,193 @@
+is connected to. Router can access space schema by using ``net.box``
+connection object contents.
+
+### When schema is used
+
+Space schema is used to flatten maps on `crud.*_object*` requests.
+Space schema is used to resolve update operation fields if updating
+by field is not supported natively. Together with sharding info,
+the space schema is used to calculate ``bucket_id`` to choose
+replicaset for a request. Finally, schema is returned to a user
+in `metadata` field of request response, so it could be used later
+for `crud.unflatten_rows`.
+
+### How schema is reloaded
+
+``net.box`` schema reload works as follows. For each request (we use ``call``s
+to execute our procedures) from client (router) to server (storage) we
+receive a schema version together with response data. If schema versions mismatch,
+client reloads schema. The request is not retried. ``net.box`` reloads
+schema before returning ``call`` result to a user, so the next request will use
+updated schema. (See [tarantool/tarantool/6169](https://github.com/tarantool/tarantool/issues/6169)).
+``crud`` cannot implicitly rely on ``net.box`` reloads: ``crud`` requests
+use ``net.box`` spaces schema to build ``net.box`` ``call`` request data,
+so if something is not relevant anymore, a part of this request data need
+to be recalculated. ``crud`` uses connection ``reload_schema`` handle
+(``ping`` wrapper, see[PR#111 comment](https://github.com/tarantool/crud/pull/111#issuecomment-765811556))
+to reload schema explicitly when needed. ``crud`` reloads each replicaset
+schema. If there are several requests to reload info, the only one reload
+fiber is started and every request waits for its completion.
+
+### When schema is reloaded and operation is retried
+
+The basic logic is as follows: if something had failed and space schema
+mismatch could be the reason, reload the schema and retry. If it didn't
+help after ``N`` retries (now ``N`` is ``1``), pass the error to the
+user.
+
+Retry with reload is triggered if
+- space not found;
+- map flatten on `crud.*_object*` request has failed;
+- ``bucket_id`` calculation has failed due to any reason;
+- operation had failed on storage, hash check is on and hashes mismatch.
+
+Let's talk a bit more about the last one. To enable hash check, the user
+should pass `add_space_schema_hash` option to a request. This option is
+always enabled for `crud.*_object*` requests. If hashes mismatch, it
+means the router schema of this space is inconsistent with the storage
+space schema, so we reload it. For ``*_many`` methods, reload and retry
+happens only if all tuples had failed with hash mismatch; otherwise,
+errors are passed to a user.
+
+Retries with reload are processed with wrappers. Wrapper is a Lua
+function that wraps the call of some business logic Lua function (for
+example, a function that transforms `crud.insert_object` map to a tuple-like
+table or a function that performs insert request with this tuple-like table).
```

Comment:
The explanation in the parentheses is redundant IMHO. I mean, the word 'wrapper' is quite usual in programming, but the explanation requires to understand several not very obvious things (what is '`crud.insert_object` map', what is 'a function that performs insert request').

Also the definition 'a wrapper is <...> that wraps' uses a term to explain itself. I would say: a wrapper of a function X is a function Y, which makes semantically same job as X, which calls X, and adds to X's execution something -- preprocessing of the inputs, postprocessing of the result, retrying. However I think it is redundant.

What is worth to say here: The reload-schema-and-retry logic is implemented as a wrapper around 'pure' `crud` request (insert, get, ...).

Reply:
Reworked